p2p/src/sync/mod.rs

```diff
@@ -488,46 +173,110 @@ where
+            // This should never happen because a peer can only connect once.
+            Some(_) => Err(P2pError::PeerError(PeerError::PeerAlreadyExists)),
+            None => Ok(()),
+        }
+    }
+
+    /// Removes the state (`PeerContext`) of the given peer.
+    fn unregister_peer(&mut self, peer: T::PeerId) {
+        log::debug!("Unregister peer {peer} from sync manager");
+
+        // Remove the queued block responses associated with the disconnected peer.
+        self.blocks_queue.retain(|(p, _)| p != &peer);
+
+        if self.peers.remove(&peer).is_some() {
+            log::warn!("Unregistering unknown peer: {peer}");
+        }
+    }
+
+    /// Announces the header of a new block to peers.
+    async fn handle_new_tip(&mut self, block_id: Id<Block>) -> Result<()> {
+        if self.is_initial_block_download {
+            self.is_initial_block_download =
+                self.chainstate_handle.call(|c| c.is_initial_block_download()).await??;
+        }
+
+        if self.is_initial_block_download {
+            return Ok(());
+        }
+
+        let header = self
             .chainstate_handle
-            .call(move |this| this.preliminary_block_check(block))
-            .await?
-        {
-            Ok(block) => {
-                self.chainstate_handle
-                    .call_mut(move |this| this.process_block(block, chainstate::BlockSource::Peer))
-                    .await?
-            }
-            Err(err) => Err(err),
-        };
-
-        let score = match result {
-            Ok(_) => 0,
-            Err(e) => match e {
-                ChainstateError::FailedToInitializeChainstate(_) => 0,
-                ChainstateError::ProcessBlockError(err) => err.ban_score(),
-                ChainstateError::FailedToReadProperty(_) => 0,
-                ChainstateError::BootstrapError(_) => 0,
-            },
-        };
-
-        if score > 0 {
-            // TODO: better abstraction over channels
-            let (tx, rx) = oneshot_nofail::channel();
-            self.tx_peer_manager
-                .send(PeerManagerEvent::AdjustPeerScore(peer_id, score, tx))
-                .map_err(P2pError::from)?;
-            let _ = rx.await.map_err(P2pError::from)?;
+            .call(move |c| c.get_block(block_id))
+            .await??
+            // This should never happen because this block has just been produced by chainstate.
+            .expect("A new tip block unavailable")
+            .header()
+            .clone();
+        self.messaging_handle.make_announcement(Announcement::Block(header))
+    }
+
+    /// Sends a block to the peer.
+    async fn send_block(&mut self, peer: T::PeerId, block: Id<Block>) -> Result<()> {
+        self.peers
+            .get_mut(&peer)
+            .ok_or(P2pError::PeerError(PeerError::PeerDoesntExist))?
+            .num_blocks_to_send
+            .checked_sub(1)
+            // This is safe because the function is called after the `blocks_queue.is_empty` check.
```

Comment:
Two things:

1. Why not add that assert at the beginning of the function?
2. Why do we even need to track the "number" if we have a queue, whose `len()` can be called?

Reply:
1. The code looks the following:
```rust
tokio::select! {
    { ... },
    (peer, block) = async { self.blocks_queue.pop_front().expect("The block queue is empty") }, if !self.blocks_queue.is_empty() => {
        let res = self.send_block(peer, block).await;
        self.handle_result(peer, res).await?;
    }
}

async fn send_block(&mut self, peer: T::PeerId, block: Id<Block>) -> Result<()> {
    self.peers
        .get_mut(&peer)
        .ok_or(P2pError::PeerError(PeerError::PeerDoesntExist))?
        .num_blocks_to_send
        .checked_sub(1)
        // This is safe because the function is called after the `blocks_queue.is_empty` check.
        .expect("Trying to send a block when num_blocks_to_send is zero");

    ...
}
```
So it is too late to assert in the `send_block` function because an item is extracted from the queue before calling it.

2. `blocks_queue` is used to store all the pending requests for every connected peer, so if there are two peers and both of them requested 400 blocks each, than the queue will contain 800 block identifiers.

I suppose this code will be change after introducing the queue for futures anyway.